python/katana/local/_graph.pyx

```diff
@@ -156,6 +188,63 @@ cdef class GraphBase:
         edge_range = self.topology().edges(n)
         return range(deref(edge_range.begin()), deref(edge_range.end()))
 
+    def get_edge_source(self, uint64_t e):
+        """
+        Get the source of an edge. This is O(log(N)) since it uses binary search to avoid having to store a source list.
+        :param e: an edge ID.
+        :return: The source node of that edge.
+        """
+        bottom = 0
+        top = self.num_nodes()
```

Comment:
Isn't there some library that already has a binary search? I am surprised still need to write it in 2021 :)

Reply:
There isn't one that I can use in this context. Cython has very poor generic programming support and doesn't really have a standard library anyway. So here we are.

I'll write a test, your concern if valid. I already shook some bugs out of it.